src/mast/mod.rs

```diff
@@ -0,0 +1,1108 @@
+    node_functions: HashMap<usize, FnInfo<B>>,
+
+    ctx: MastCtx,
+
+    /// Monomorphized AST of the main function
+    pub main_fn_ast: Option<FunctionDef>,
+}
+
+impl<B: Backend> Mast<B> {
+    pub fn new(tast: TypeChecker<B>) -> Self {
+        Self {
+            tast,
+            node_types: HashMap::new(),
+            node_functions: HashMap::new(),
+            ctx: MastCtx::new(),
+            main_fn_ast: None,
+        }
+    }
+
+    pub fn expr_type(&self, expr: &Expr) -> Option<&TyKind> {
+        self.node_types.get(&expr.node_id)
+    }
+
+    pub fn expr_fn(&self, expr: &Expr) -> Option<&FnInfo<B>> {
+        self.node_functions.get(&expr.node_id)
+    }
+
+    pub fn struct_info(&self, qualified: &FullyQualified) -> Option<&StructInfo> {
+        self.tast.struct_info(qualified)
+    }
+
+    pub fn const_info(&self, qualified: &FullyQualified) -> Option<&ConstInfo<B::Field>> {
+        self.tast.const_info(qualified)
+    }
+
+    pub fn size_of(&self, typ: &TyKind) -> usize {
+        self.tast.size_of(typ)
+    }
+
+    /// Monomorphize the main function.
+    /// This is the entry point of the monomorphization process.
+    /// It stores the monomorphized AST at the end.
+    pub fn monomorphize(&mut self) -> Result<()> {
+        // store mtype in mast
+        let qualified = FullyQualified::local("main".to_string());
+        let main_fn = self
+            .tast
+            .fn_info(&qualified)
+            .ok_or(self.error(ErrorKind::NoMainFunction, Span::default()))?;
+
+        let func_def = match &main_fn.kind {
+            // `fn main() { ... }`
+            FnKind::Native(function) => function.clone(),
+
+            _ => panic!("main function must be native"),
+        };
+
+        // create a new typed fn environment to type check the function
+        let mut typed_fn_env = MonomorphizedFnEnv::default();
+
+        // store variables and their types in the fn_env
+        for arg in &func_def.sig.arguments {
+            // store the args' type in the fn environment
+            typed_fn_env.store_type(
+                &arg.name.value,
+                &MTypeInfo::new(&arg.typ.kind, arg.span, None),
+            )?;
+        }
+
+        // the output value returned by the main function is also a main_args with a special name (public_output)
+        if let Some(typ) = &func_def.sig.return_type {
+            match typ.kind {
+                TyKind::Field => {
+                    typed_fn_env.store_type(
+                        "public_output".as_ref(),
+                        &MTypeInfo::new(&typ.kind, typ.span, None),
+                    )?;
+                }
+                TyKind::Array(_, _) => {
+                    typed_fn_env.store_type(
+                        "public_output".as_ref(),
+                        &MTypeInfo::new(&typ.kind, typ.span, None),
+                    )?;
+                }
+                _ => unimplemented!(),
+            }
+        }
```

Comment:
I think it is intended to propagate the types from the arguments. Otherwise, the expressions depend on these arguments can't be computed for their type.